src/Npgsql/NpgsqlConnector.cs

```diff
@@ -450,9 +450,74 @@ void WriteStartupMessage(string username)
             if (timezone != null)
                 startupParams["TimeZone"] = timezone;
 
+            var options = Settings.Options?.Length > 0
```

Comment:
How about moving all the parsing logic to NpgsqlConnectionStringBuilder, and exposing a `Dictionary<string,string>` property there? This would encapsulate all of the connection string parsing logic within that class.

Reply:
Yeah I briefly thought about this but than discarded it because it wouldn't work sanely for environment variables that way.
Now that you suggested it again, I thought about it once more and came to the conclusion that, while caching the parsed value from the connection string is good, caching of the parsed value of the environment variable might even be problematic because the user might expect the application to immediately honor changed environment variables, so this is probably even what we want.

I implemented it that way now.

Reply:
I don't think it's technically possible change environment variables from outside the process once the process has started - at least in Unix, the moment a process is started, they inherit the environment variables from their parent and from that point they have two distinct sets of values. The process can change the environment for themselves, but that's a bit strange.

Anyway, it's not a big deal either way.

Reply:
Ok, I've added transparent caching of options from environment variables too.